docs/upload_specifications.ADR.md

```diff
@@ -0,0 +1,170 @@
+# Upload Specifications Architecture Design Record
+
+This document specifies the design for handling upload specifications in the Staging Service (StS).
+An upload specification is an Excel, CSV, or TSV file that contains instructions for how
+to upload one or more files in the staging area to KBase as KBase data types.
+
+## Resources
+
+* [The original strategy document for this approach](https://docs.google.com/document/d/1ocmZVBlTzAh_cdZaWGRwIbAuH-mcPRZFdhcwRhAfzxM/edit)
+  * Readable for everyone that has access to the KBase Google Docs folder
+  * The document contains short descriptions of future projects not included in this work,
+    such as generating template files on the fly and rich templates.
+
+## Front end changes
+
+The design introduces a new StS data type, `upload_specification`. The FE's current 
+behavior is to display any data types in the file dropdown, but ignore files for which
+the selected data type is unknown, a bug. The FE will be updated to ignore unknown
+data types returned from the StS, allowing for phased, non-lockstep upgrades. This work is not
+included in this project, but will be in a future FE project.
+
+## Upload specification input files
+
+Input file formats may be Excel, CSV, or TSV. An example CSV file structure for GFF/FASTA uploads
+is below:
+
+```
+Data type:, gff_metagenome
```

Comment:
https://csvlint.io/validation/610334f0a9b69a000400003c had some comments on this, not sure if its useful.

There might be multiple valid apps for a given datatype. If the user is specifying the app, then that might be a better route. We could also include that info  in the header of the CSV as well. 

We could get rid of the first line, and have that information container in the suffix of the file, like "Project1.soil.gff_metagenome.up_spec"
And then have the next two lines merged into one as the header line of the CSV, so you have a header and the rest are rows.
For excel files, we could use the sheet titles as the way to specify the apps/datatypes.

Reply:
Re the CSV validation - it's saying the CSV file has 13 rows, which is way off. There seems to be something wrong there.

> There might be multiple valid apps for a given datatype. 
I don't think this is possible - there has to be a 1:1 mapping from data type to app for the bulk cell to display itself.

> We could get rid of the first line, and have that information container in the suffix of the file
Way better to have the data travel with the file contents than the name, which can be changed IMO.

> And then have the next two lines merged into one as the header line of the CSV, so you have a header and the rest are rows.
That makes it a lot harder for humans to read, IMO.

Reply:
What about the FASTQ or GFF Uploaders? Aren't those not a 1:1 mapping?

Are we forcing specific delimiters or newlines or quotechars for these character limited files? (Does CSV stand for comma or character)?

Do we care about having multiple header lines? I've never really come across any CSVs that do that.

Reply:
> What about the FASTQ or GFF Uploaders? Aren't those not a 1:1 mapping?

Not on a data type basis. gff_metagenome, gff_genome, fastq_reads_interleaved, and fastq_reads_noninterleaved are all 1:1.

On a *file type* basis yes, but that's not the type on the line in question.

> Are we forcing specific delimiters or newlines or quotechars for these character limited files? (Does CSV stand for comma or character)?

CSV is comma separated values, so yeah on forcing delimiters. We shouldn't have to force newlines since python handles either. For quoting, I'll just figure out whatever Pandas supports and then document that in the readme (https://kbase-jira.atlassian.net/browse/DATAUP-547)

> Do we care about having multiple header lines?

I don't, and I don't see a reason that should be a problem as long as the parser expects it.

Reply:
Well, maybe Pandas, maybe just a CSV reader. Let's not be hasty in specifying the implemenation